sw/device/lib/dif/dif_hmac.h

```diff
@@ -0,0 +1,211 @@
+ */
+typedef enum dif_hmac_error {
+  /** No error occurred. */
+  kDifHmacNoError = 0,
+  /** An invalid argument was provided. */
+  kDifHmacInvalidArgument,
+} dif_hmac_error_t;
+
+/**
+ * Error codes for HMAC FIFO operations that may fail.
+ */
+typedef enum dif_hmac_fifo_error {
+  /** No error occurred. */
+  kDifHmacFifoNoError = 0,
+  /** An invalid argument was provided. */
+  kDifHmacFifoInvalidArgument,
+  /** The FIFO filled up before the buffer was fully consumed. */
+  kDifHmacFifoFull,
+} dif_hmac_fifo_error_t;
+
+/**
+ * Configuration for initializing the HMAC device.
+ */
+typedef struct dif_hmac_config {
+  /** The base address for registers in the HMAC IP*/
+  mmio_region_t base_addr;
+  /** The mode of operation */
+  dif_hmac_mode_t mode;
+  /** Byte endianness for writes to the FIFO */
+  dif_hmac_endianness_t message_endianness;
+  /** Byte endianness for reads from the digest */
+  dif_hmac_endianness_t digest_endianness;
+  /** The HMAC key. Only used in |kDifHmacModeHmac| mode. */
+  uint32_t key[8];
+} dif_hmac_config_t;
+
+/**
+ * A typed representation of the HMAC digest.
+ */
+typedef struct dif_hmac_digest {
+  uint32_t digest[8];
+} dif_hmac_digest_t;
+
+/**
+ * State for a particular HMAC device.
+ *
+ * its member variables should be considered private, and are only provided so
+ * that callers can allocate it.
+ */
+typedef struct dif_hmac {
+  mmio_region_t base_addr;
+} dif_hmac_t;
+
+/**
+ * Initializes the HMAC device described by |config|, writing internal state to
+ * |hmac_out|.
+ *
+ * This function *must* be called on a particular |mmio_region_t| before calling
+ * any other functions in this header with that |mmio_region_t|; moreover, it
+ * *may not* be called more than once with any |mmio_region_t|. All deviations
+ * are Undefined Behavior.
+ *
+ * @param config Configuration supplied for initializing a particular device.
+ * @param hmac_out The location at which to write HMAC state. This location
+ *                 must be valid to write to.
+ * @return |kHmacInvalidArgument| if |config| is null or contains illegal
+ *          arguments or |hmac_out| is null, |kHmacNoError| otherwise.
+ */
+dif_hmac_error_t dif_hmac_init(const dif_hmac_config_t *config,
+                               dif_hmac_t *hmac_out);
+
+/**
+ * Attempts to send |len| bytes from the buffer pointed to by |data| to the
+ * device described by |hmac|. This function will send to the message FIFO until
+ * the FIFO fills up or |len| bytes have been sent.
+ *
+ * In the event that the FIFO fills up before |len| bytes have been sent this
+ * function will return a |kDifHmacFifoFull| error.
+ *
+ * |data| *must* point to an allocated buffer of at least length |len|.
+ *
+ * @param hmac The HMAC device to send to.
+ * @param data A contiguous buffer to copy from.
+ * @param len The length of the buffer to copy from.
+ * @param The number of bytes sent to the FIFO (optional).
```

Comment:
`@param bytes_sent ...`

This discussion is outside of the scope of this review, but:
I really think it would be nice to have a tool that just verifies that function signatures match the function description, I suggested that to @asb and @lenary , just something to think about.